File: js/widgets/slider.ts
import type { RenderProps } from '@anywidget/types';

import { Slider } from 'https://cdn.jsdelivr.net/npm/@uwdata/mosaic-inputs@0.16.2/+esm';

import { vizCoordinator } from '../coordinator';

interface SliderProps {
    df_id: string;
    column: string;
    param: string;
    params: string;
}

async function render({ model, el }: RenderProps<SliderProps>) {
    // unwrap widget parameters
    const df_id: string = model.get('df_id');
    const column: string = model.get('column');
    const param: string = model.get('param');

    // get the data frame
    const coordinator = await vizCoordinator();
    await coordinator.waitForData(df_id);

    // add params
    coordinator.addParams(JSON.parse(model.get('params')));

    // initialize the slider and connect it
    const menu = new Slider({
        element: el,
        as: coordinator.getParam(param),
        from: df_id,
        column: column,
    });
    await coordinator.connectClient(menu);
}

export default { render };

File: src/inspect_viz/_inputs/slider.py
import traitlets
from anywidget import AnyWidget

from .._data import Data, validate_bindings
from .._param import Param
from .._util.constants import STATIC_DIR


def slider(df: Data, column: str, param: Param | None = None) -> AnyWidget:
    validate_bindings(df, column, param)

    class SliderInput(AnyWidget):
        _esm = STATIC_DIR / "slider.js"
        df_id = traitlets.CUnicode("").tag(sync=True)
        column = traitlets.CUnicode("").tag(sync=True)
        param = traitlets.CUnicode("").tag(sync=True)
        params = traitlets.CUnicode(Param.get_all_as_json()).tag(sync=True)

    return SliderInput(df_id=df.id, column=column, param=param.id if param else "")
